enhancements/network/egress-qos.md

```diff
@@ -0,0 +1,209 @@
+The CRD will be Namespaced, with one resource allowed per namespace.
+The resources will be watched by ovn-k, which in turn will configure OVN's [QoS Table](https://man7.org/linux/man-pages/man5/ovn-nb.5.html#QoS_TABLE). 
+
+## Motivation
+
+Telco customers require support for DSCP marking capability for some of their 5G applications, giving some pods precedence over others.
+The QoS markings will be consumed and acted upon by objects outside of the OpenShift cluster to optimize traffic flow throughout their networks.
+
+### Goals
+
+- Provide a mechanism for users to set DSCP on egress traffic coming from specific namespaces.
+
+### Non-Goals
+
+- Ingress QoS.
+
+- Consolidating with current `kubernetes.io/egress-bandwidth` and `kubernetes.io/ingress-bandwidth` annotations.
+Nonetheless, the work done here does not interfere with the current bandwidth QoS mechanism.
+
+- The DSCP marking does not need to be handled or acted upon by OpenShift, just added to selected headers.
+
+- Marking East/West traffic, exposing the DSCP value from the inner packet to the outer geneve packet.
+
+## Proposal
+
+To achieve egress DSCP marking on pods, we introduce a new namespace-scoped CRD `EgressQoS` which allows specifying a set of QoS rules, each has a DSCP value, a destination CIDR and a PodSelector.
+Traffic coming from pods on the namespace heading to each destination CIDR that match the selector will be marked with the corresponding DSCP value.
+Not specifying a PodSelector will apply a rule to all pods in the namespace.
+
+### Implementation Details/Notes/Constraints
+
+A new API `EgressQoS` under the `k8s.ovn.org/v1` version will be added to `pkg/crd`.
+
+A new controller in OVN-K will watch `EgressQoS`, `Pod` and `Node` objects, which will create the relevant QoS objects in OVN and result in the necessary flows to be programmed in OVS. To which logical switches these QoS objects are attached to differs a bit between SGW/LGW modes:
+* In SGW mode it is enough to attach the QoS objects only to the single distributed `join` switch, as all pods egress traffic goes through it.
+* In LGW mode we need to attach the QoS objects to all of the node local switches in the cluster.
+
+In order to not create an OVN QoS object per pod in the namespace, the controller will also manage AddressSets. 
+For each QoS rule specified in a given `EgressQoS` it'll create an AddressSet, adding only the pods
+whose label matches the PodSelector to it, making sure that new/updated/deleted matching pods are also added/updated/deleted accordingly.
+Rules that do not have a PodSelector will leverage the namespace's AddressSet.
+
+For example, using LGW mode and assuming there's a single node `node1` and the following `EgressQoS` is created:
+
+```yaml
+kind: EgressQoS
+apiVersion: k8s.ovn.org/v1
+metadata:
+  name: default
+  namespace: default
+spec:
+  egress:
+  - dscp: 46
+    dstCIDR: 1.2.3.4/32
+    podSelector:
+      matchLabels:
+        priority: Critical
```

Comment:
What does the `priority: Critical` mean in the context of the `podSelector`?

Reply:
Oh it is just an example of matching pods with labels as critical?

Reply:
yes :sweat_smile: